runtime/src/accounts_index.rs

```diff
@@ -181,9 +201,169 @@ impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
+            // the previous index entry.
+            if prev_key != *key {
+                self.remove_index_entries(&prev_key, inner_key, &[slot]);
+            }
+        }
+    }
+
+    fn remove_index_entries(&self, key: &Pubkey, inner_key: &Pubkey, slots: &[Slot]) {
+        let is_key_empty = if let Some(inner_key_map) = self.index.get(&key) {
+            // Delete the slot from the slot set
+            let is_inner_key_empty = if let Some(slot_set) = inner_key_map.get(&inner_key) {
+                for slot in slots.iter() {
+                    if slot_set.remove(slot).is_none() {
+                        warn!("Reverse index is missing previous entry for key {}, inner_key: {}, slot: {}",
+                        key, inner_key, slot);
+                    }
+                }
+
+                slot_set.is_empty()
+            } else {
+                false
+            };
+
+            // Check if `key` is empty
+            let is_key_empty = if is_inner_key_empty {
+                if let dashmap::mapref::entry::Entry::Occupied(inner_key_entry) =
+                    inner_key_map.entry(*inner_key)
+                {
+                    // Delete the `inner_key` if the slot set is empty
+                    let slot_set = inner_key_entry.get();
+                    if slot_set.is_empty() {
+                        inner_key_entry.remove();
+                    }
+                }
+                inner_key_map.is_empty()
+            } else {
+                false
+            };
+            is_key_empty
+        } else {
+            false
+        };
+
+        // Delete the `key` if the set of inner keys is epty
+        if is_key_empty {
+            if let dashmap::mapref::entry::Entry::Occupied(key_entry) = self.index.entry(*key) {
+                if key_entry.get().is_empty() {
+                    key_entry.remove();
+                }
+            }
+        }
+    }
+
+    // Specifying `slots_to_remove` == None will only remove keys for those specific slots
+    // found for the `inner_key` in the reverse index. Otherwise, passing `None`
+    // will  remove all keys that are found for the `inner_key` in the reverse index.
+    fn remove_by_inner_key(&self, inner_key: &Pubkey, slots_to_remove: Option<&HashSet<Slot>>) {
+        // Save off which keys in `self.index` had slots removed so we can remove them
+        // after we purge the reverse index
+        let mut key_to_removed_slots: HashMap<Pubkey, Vec<Slot>> = HashMap::new();
+
+        // Check if the entry for `inner_key` in the reverse index is empty
+        // and can be removed
+        let needs_remove = {
+            if let Some(slots_to_remove) = slots_to_remove {
+                self.reverse_index
+                    .get(inner_key)
+                    .map(|slots_map| {
+                        for slot in slots_to_remove.iter() {
+                            if let Some(removed) = slots_map.value().remove(slot) {
```

Comment:
@ryoqun  I think this is safe because when `slots_to_remove == None` is passhere:  https://github.com/solana-labs/solana/blob/f4f906a9531d1ff9565996cd8848c712a58d9f68/runtime/src/accounts_index.rs#L669, so it shouldn't clear any other thread's writes to this slot's secondary index, the accounts index lock for the slot is held here: https://github.com/solana-labs/solana/blob/f4f906a9531d1ff9565996cd8848c712a58d9f68/runtime/src/accounts_index.rs#L666